### ISSUE TITLE
HalfTensor Training Needs non-Stateless Method in F.Linear

### ISSUE THREAD

**Issue by ajbrock** [pytorch/pytorch#827]
Hi Guys,

Currently trying to call backward() on a network trained in FP16 with a linear output results in an error:

```
import torch
import torch.nn.functional as F
from torch.autograd import Variable as V
x = V(torch.randn(20,5).cuda().half(),requires_grad=True)
inp = V(torch.randn(100,5).cuda().half())
out = F.linear(inp,x)
out2 = out.mean()
out2
Variable containing:
1.00000e-02 *
 -4.1840
[torch.cuda.HalfTensor of size 1 (GPU 0)]


out2.backward()
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/visionlab/anaconda2/lib/python2.7/site-packages/torch/autograd/variable.py", line 145, in backward
    self._execution_engine.run_backward((self,), (gradient,), retain_variables)
  File "/home/visionlab/anaconda2/lib/python2.7/site-packages/torch/nn/_functions/linear.py", line 25, in backward
    grad_weight = torch.mm(grad_output.t(), input)
RuntimeError: Type HalfTensor doesn't implement stateless methods
```

This is easily fixed by replacing these calls with the non-stateless versions, as I've done [here](https://github.com/ajbrock/pytorch/blob/master/torch/nn/_functions/linear.py). I'm not sure if this replacement is in line with pytorch style (or if there are plans to implement stateless methods for HalfTensors which would render this irrelevant), but I can submit a PR if needbe.

Best,

Andy

**Comment by apaszke** [comment]
If it has an mm method then we should enable `torch.mm` for it too. It should be a one line change in cwrap.

**Comment by soumith** [comment]
@apaszke we dont enable stateless methods for any half tensors at the moment.

**Comment by apaszke** [comment]
Fixed in #852.
